***@***.***

chaiwat

ในวันที่ ศ. 17 ก.ย. 2021 18:31 Jon Atack ***@***.***>
เขียนว่า:

> ***@***.**** commented on this pull request.
>
> Tentative concept ACK after looking at the callers of GetFee() (not
> GetFeePerK, which should be a pass-through). The new functional tests
> fail where expected on master.
>
> $ test/functional/rpc_fundrawtransaction.py
>
>
>
>   File "/home/jon/projects/bitcoin/bitcoin/test/functional/rpc_fundrawtransaction.py", line 1055, in test_feerate_rounding
>
>     assert_raises_rpc_error(-4, "Insufficient funds", w.fundrawtransaction, rawtx, {"fee_rate": 1.85})
>
> AssertionError: Unexpected stderr bitcoind: wallet/spend.cpp:814: bool CreateTransactionInternal(CWallet &, const std::vector<CRecipient> &, CTransactionRef &, CAmount &, int &, bilingual_str &, const CCoinControl &, FeeCalculation &, bool): Assertion `coin_selection_params.m_subtract_fee_outputs || fee_needed <= change_and_fee - change_amount' failed. !=
>
>
> $ test/functional/wallet_keypool.py
>
>
>
>   File "/home/jon/projects/bitcoin/bitcoin/test/functional/wallet_keypool.py", line 184, in run_test
>
>     assert_equal(res["fee"], Decimal("0.00009706"))
>
>   File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 49, in assert_equal
>
>     raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
>
> AssertionError: not(0.00009705 == 0.00009706)
>
>
> ------------------------------
>
> In src/policy/feerate.cpp
> <https://github.com/bitcoin/bitcoin/pull/22949#discussion_r710958422>:
>
> > @@ -22,7 +24,7 @@ CAmount CFeeRate::GetFee(uint32_t num_bytes) const
>
>  {
>
>      const int64_t nSize{num_bytes};
>
>
>
> -    CAmount nFee = nSatoshisPerK * nSize / 1000;
>
> +    CAmount nFee = std::ceil(nSatoshisPerK * nSize / 1000.0);
>
>
> ee9a74d
> <https://github.com/bitcoin/bitcoin/commit/ee9a74d99ce784b3570ae606eac4dd9627727ec1>
> a couple of ideas
>
>    - explicit, named cast
>
> ⬇️ Suggested change
>
> -    CAmount nFee = std::ceil(nSatoshisPerK * nSize / 1000.0);
>
> +    // Be explicit that we're converting from double to long (CAmount) here.
>
> +    // We've previously had bugs creep in from silent double->int conversion.
>
> +    CAmount nFee{static_cast<CAmount>(std::ceil(nSatoshisPerK * nSize / 1000.0))};
>
>
> ------------------------------
>
>    - bypass this calculation in the frequent GetFeePerK() case
>
>     if (nSize != 1000) {
>
>         CAmount nFee{static_cast<CAmount>(std::ceil(nSatoshisPerK * nSize / 1000.0))};
>
>     }
>
> ------------------------------
>
> In test/functional/rpc_fundrawtransaction.py
> <https://github.com/bitcoin/bitcoin/pull/22949#discussion_r710959094>:
>
> > +
>
> +        addr = w.getnewaddress(address_type="bech32")
>
> +        self.nodes[0].sendtoaddress(addr, 1)
>
> +        self.generate(self.nodes[0], 1)
>
> +        self.sync_all()
>
> +
>
> +        # A P2WPKH input costs 68 vbytes; With a single P2WPKH output, the rest of the tx is 42 vbytes for a total of 110 vbytes.
>
> +        # At a feerate of 1.85 sat/vb, the input will need a fee of 125.8 sats and the rest 77.7 sats
>
> +        # The entire tx fee should be 203.5 sats.
>
> +        # Coin selection rounds the fee individually instead of at the end (due to how CFeeRate::GetFee works).
>
> +        # If rounding down (which is the incorrect behavior), then the calculated fee will be 125 + 77 = 202.
>
> +        # If rounding up, then the calculated fee will be 126 + 78 = 204.
>
> +        # In the former case, the calculated needed fee is higher than the actual fee being paid, so an assertion is reached
>
> +        # To test this does not happen, we subtract 202 sats from the input value. If working correctly, this should
>
> +        # fail with insufficient funds rather than bitcoind asserting.
>
> +        rawtx = w.createrawtransaction([], [{self.nodes[0].getnewaddress(address_type="bech32"): 1 - 0.00000202}])
>
>
> 053ec2b
> <https://github.com/bitcoin/bitcoin/commit/053ec2b901bee174de3243302c113b4a91733241>
> nit
> ⬇️ Suggested change
>
> -        rawtx = w.createrawtransaction([], [{self.nodes[0].getnewaddress(address_type="bech32"): 1 - 0.00000202}])
>
> +        rawtx = w.createrawtransaction(inputs=[], outputs=[{self.nodes[0].getnewaddress(address_type="bech32"): 1 - 0.00000201}])
>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/22949#pullrequestreview-757331938>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APA4O7QPOSWJU3L3MRHTVOLUCMRI7ANCNFSM5D2K6V6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

--- a/test/functional/test_framework/util.py
+++ b/test/functional/test_framework/util.py
@@ -36,12 +36,12 @@ def assert_approx(v, vexp, vspan=0.00001):
 
 def assert_fee_amount(fee, tx_size, feerate_BTC_kvB):
     """Assert the fee is in range."""
-    feerate_BTC_vB = feerate_BTC_kvB / 1000
-    target_fee = satoshi_round(tx_size * feerate_BTC_vB)
+    target_fee = get_fee(tx_size, feerate_BTC_kvB)
     if fee < target_fee:
         raise AssertionError("Fee of %s BTC too low! (Should be %s BTC)" % (str(fee), str(target_fee)))
     # allow the wallet's estimation to be at most 2 bytes off
-    if fee > (tx_size + 2) * feerate_BTC_vB:
+    high_fee = get_fee(tx_size + 2, feerate_BTC_kvB)
+    if fee > high_fee:
         raise AssertionError("Fee of %s BTC too high! (Should be %s BTC)" % (str(fee), str(target_fee)))
 
 
@@ -216,6 +216,18 @@ def count_bytes(hex_string):
 
 def str_to_b64str(string):
     return b64encode(string.encode('utf-8')).decode('ascii')
+
+
+def ceildiv(a, b):
+    """Divide 2 ints and round up to next int rather than round down"""
+    return -(-a // b)
+
+
+def get_fee(tx_size, feerate_btc_kvb):
+    """Calculate the fee in BTC given a feerate is BTC/kvB. Reflects CFeeRate::GetFee"""
+    feerate_sat_kvb = int(feerate_btc_kvb * Decimal(1e8)) # Fee in sat/kvb as an int to avoid float precision errors
+    target_fee_sat = ceildiv(feerate_sat_kvb * tx_size, 1000) # Round calculated fee up to nearest sat
+    return satoshi_round(target_fee_sat / Decimal(1e8)) # Truncate BTC result to nearest sat
 
 
 def satoshi_round(amount):

